### PR TITLE
Rule Manager: Enhance load hotspot rules

### PR DESCRIPTION
**Pull request by liqiangz**
### Describe what this PR does / why we need it
Enhance load hotspot rules

### Describe how you did it

1.Use `copy on update` to reduce write lock.
2.Use updateLock serialize loading rules.

Existing logic: lock.Lock() -> build new tcMap by old tcMap -> set tcMap ->lock.Unlock()

New logic: updateLock.lock() -> lock.RLock() -> build clone tcMap -> lock.RUnlock -> build new tcMap by clone tcMap -> lock.Lock() -> set tcMap -> lock.Unlock () -> updateLock.Unlock()

### Special notes for reviews

### CODE REVIEW

**Comment by louyuting** [comment]
@liqiangz   Could you please describe your enhancement detail in PR?

**Comment by liqiangz** [comment]
@louyuting   I have added some details

File: core/hotspot/rule_manager.go [hunk 1]
```diff
@@ -219,6 +223,16 @@ func onRuleUpdate(rules []*Rule) (err error) {
 			}
 		}
 	}
+
+	tcMux.Lock()
+	defer func() {
+		tcMux.Unlock()
+		if r := recover(); r != nil {
+			return
+		}
```
Comment (luckyxiaoqiang) [inline core/hotspot/rule_manager.go:232]: Seems no need to do recover at here, as there are only some very simple assignment operations.

File: core/hotspot/rule_manager.go [hunk 2]
```diff
@@ -72,6 +73,8 @@ func getTrafficControllersFor(res string) []TrafficShapingController {
 // bool: indicates whether the internal map has been changed;
 // error: indicates whether occurs the error.
 func LoadRules(rules []*Rule) (bool, error) {
+	updateRuleMux.Lock()
+	defer updateRuleMux.Unlock()
 	tcMux.RLock()
```
Comment (sanxun0325) [inline core/hotspot/rule_manager.go:78]: Perhaps tcMux.Rlock is not necessary; currentRules can be locked with updateRuleMux

Comment (liqiangz) [reply core/hotspot/rule_manager.go]: Done, thanks.

File: core/hotspot/rule_manager.go (hunk 1 above)
Comment (liqiangz) [reply core/hotspot/rule_manager.go]: Done, thanks.

**Review by sanxun0325 — DISMISSED** [review]
LGTM

File: core/hotspot/rule_manager.go [hunk 3]
```diff
@@ -219,6 +221,13 @@ func onRuleUpdate(rules []*Rule) (err error) {
 			}
 		}
 	}
+
+	tcMux.Lock()
+	defer func() {
```
Comment (louyuting) [inline core/hotspot/rule_manager.go:226]: remove defer and move logging logic to the first defer

**Review by louyuting — CHANGES_REQUESTED** [review]

File: core/hotspot/rule_manager.go (hunk 3 above)
Comment (liqiangz) [reply core/hotspot/rule_manager.go]: done

**Review by louyuting — DISMISSED** [review]

File: core/hotspot/rule_manager.go [hunk 4]
```diff
@@ -219,18 +220,29 @@ func onRuleUpdate(rules []*Rule) (err error) {
...
+	logging.Debug("[HotSpot onRuleUpdate] Time statistic(ns) for updating hotSpot rule", "timeCost", util.CurrentTimeNano()-start)
+	logRuleUpdate(newRuleMap)
 	return nil
 }
 
-func logRuleUpdate(m trafficControllerMap) {
-	rs := rulesFrom(m)
-	if len(rs) == 0 {
+func logRuleUpdate(m map[string][]*Rule) {
+	rules := make([]*Rule, 0, 9)
```
Comment (louyuting) [inline core/hotspot/rule_manager.go:235]: slice length changes back to 8.

Comment (liqiangz) [reply core/hotspot/rule_manager.go]: done

**Review by louyuting — APPROVED** [review]

**Pull request merged**